File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class TreeNode {
    char val;
    boolean eow;
    Map<Character, TreeNode> subs;

    TreeNode(char val) {
        this.val = val;
        this.subs = new HashMap<>();
        this.eow = false;
    }

    TreeNode() {
        this.subs = new HashMap<>();
        this.eow = false;
    }

    public TreeNode add(char ch) {
        TreeNode next = new TreeNode(ch);
        subs.putIfAbsent(ch, next);
        return next;
    }
}

class Trie {
    TreeNode root;

    public Trie() {
        root = new TreeNode();
    }
    
    public void insert(String word) {
        TreeNode cur = root;

        for (char ch: word.toCharArray()) {
            if(!cur.subs.containsKey(ch)) {
                cur = cur.add(ch);
            } else {
                cur = cur.subs.get(ch);
            }
        }

        cur.eow = true;
    }
    
    public boolean search(String word) {
        TreeNode cur = root;
        for (char ch: word.toCharArray()) {
            if(!cur.subs.containsKey(ch)) {
                return false;
            }

            cur = cur.subs.get(ch);
        }

        return cur.eow;
    }
    
    public boolean startsWith(String prefix) {
        TreeNode cur = root;
        for (char ch: prefix.toCharArray()) {
            if(!cur.subs.containsKey(ch)) {
                return false;
            }

            cur = cur.subs.get(ch);
        }

        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.kt
class TrieNode {
    val links: Array<TrieNode?>
    var isEnd: Boolean = false
    
    constructor() {
        links = Array<TrieNode?>(26){ null }
    }
    
    fun has(ch: Char): Boolean {
        return links[ch - 'a'] != null
    }
    
    fun get(ch: Char): TrieNode? {
        return links[ch - 'a']
    }
    
    fun put(ch: Char, node: TrieNode) {
        links[ch - 'a'] = node
    }
}

class Trie() {
    
    val root = TrieNode();

    fun insert(word: String) {
        var curr: TrieNode? = root
        
        for(ch in word) {
            if (!curr!!.has(ch)) {
                curr.put(ch, TrieNode())
            }
            
            curr = curr.get(ch)
        }
        
        curr!!.isEnd = true
    }

    fun search(word: String): Boolean {
        var curr: TrieNode? = root
        
        for (ch in word) {
            if (!curr!!.has(ch)) return false
            curr = curr!!.get(ch)
        }
        
        return curr!!.isEnd
    }

    fun startsWith(prefix: String): Boolean {
        var curr: TrieNode? = root
        
        for (ch in prefix) {
            if (!curr!!.has(ch)) return false
            curr = curr!!.get(ch)
        }
        
        return true
    }

}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */